ext/crates/fp/src/vector/mod.rs

```diff
@@ -1,491 +1,39 @@
-        pub fn add_unmasked(&mut self, other: Slice, c: u32, mask: &[usize]);
-    }
-
-    pub fn add_tensor(&mut self, offset: usize, coeff: u32, left: Slice, right: Slice) {
-        match (self, left, right) {
-            (SliceMut::_2(x), Slice::_2(y), Slice::_2(z)) => x.add_tensor(offset, coeff, y, z),
-            (SliceMut::_3(x), Slice::_3(y), Slice::_3(z)) => x.add_tensor(offset, coeff, y, z),
-            (SliceMut::_5(x), Slice::_5(y), Slice::_5(z)) => x.add_tensor(offset, coeff, y, z),
-            (SliceMut::_7(x), Slice::_7(y), Slice::_7(z)) => x.add_tensor(offset, coeff, y, z),
-            _ => {
-                panic!("Applying add_tensor to vectors over different primes");
-            }
-        }
-    }
-}
-
-impl<'a> FpVectorNonZeroIterator<'a> {
-    dispatch_vector! {
-        fn next(&mut self) -> (Option<(usize, u32)>);
-    }
-}
-
-impl std::fmt::Display for FpVector {
-    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
-        self.as_slice().fmt(f)
-    }
-}
-
-impl<'a> std::fmt::Display for Slice<'a> {
-    /// # Example
-    /// ```
-    /// # use fp::vector::FpVector;
-    /// # use fp::prime::ValidPrime;
-    /// let v = FpVector::from_slice(ValidPrime::new(2), &[0, 1, 0]);
-    /// assert_eq!(&format!("{v}"), "[0, 1, 0]");
-    /// assert_eq!(&format!("{v:#}"), "010");
-    /// ```
-    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
-        if f.alternate() {
-            for v in self.iter() {
-                write!(f, "{v}")?;
-            }
-            Ok(())
-        } else {
-            write!(f, "[{}]", self.iter().format(", "))
-        }
-    }
-}
-
-impl From<&FpVector> for Vec<u32> {
-    fn from(v: &FpVector) -> Vec<u32> {
-        v.iter().collect()
-    }
-}
-
-impl std::ops::AddAssign<&FpVector> for FpVector {
-    fn add_assign(&mut self, other: &FpVector) {
-        self.add(other, 1);
-    }
-}
-
-impl<'a> Iterator for FpVectorNonZeroIterator<'a> {
-    type Item = (usize, u32);
-
-    fn next(&mut self) -> Option<Self::Item> {
-        self.next()
-    }
-}
-
-impl<'a> IntoIterator for &'a FpVector {
-    type IntoIter = FpVectorIterator<'a>;
-    type Item = u32;
-
-    fn into_iter(self) -> Self::IntoIter {
-        self.iter()
-    }
-}
-
-macro_rules! impl_try_into {
-    ($var:tt, $p:literal) => {
-        impl<'a> TryInto<&'a mut FpVectorP<$p>> for &'a mut FpVector {
-            type Error = ();
-
-            fn try_into(self) -> Result<&'a mut FpVectorP<$p>, ()> {
-                match self {
-                    FpVector::$var(x) => Ok(x),
-                    _ => Err(()),
-                }
-            }
-        }
+pub mod base_generic;
+pub mod generic;
+pub(crate) mod internal;
+
+#[cfg(feature = "odd-primes")]
+pub mod specialized;
+pub mod specialized_2;
+
+pub use specialized::{FpVector, FpVectorNonZeroIterator, Slice, SliceMut};
+#[cfg(not(feature = "odd-primes"))]
+pub use specialized_2 as specialized;
```

Comment:
I think you want to swap the order of these two statements.

Reply:
Wow, yeah I don't know how I didn't catch that before